runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/jobsubmission/PortablePipelineJarCreator.java

```diff
@@ -0,0 +1,289 @@
+    Preconditions.checkArgument(
+        classPathResources.size() == 1, "Expected exactly one jar on " + classLoader.toString());
+    copyResourcesFromJar(new JarFile(classPathResources.get(0)), outputStream);
+  }
+
+  /** Copy resources from {@code inputJar} to {@code outputStream}. */
+  @VisibleForTesting
+  protected static void copyResourcesFromJar(JarFile inputJar, JarOutputStream outputStream)
+      throws IOException {
+    Enumeration<JarEntry> inputJarEntries = inputJar.entries();
+    // The zip spec allows multiple files with the same name; the Java zip libraries do not.
+    // Keep track of the files we've already written to filter out duplicates.
+    // Also, ignore the old manifest; we want to write our own.
+    Set<String> previousEntryNames = new HashSet<>(ImmutableList.of(JarFile.MANIFEST_NAME));
+    while (inputJarEntries.hasMoreElements()) {
+      JarEntry inputJarEntry = inputJarEntries.nextElement();
+      InputStream inputStream = inputJar.getInputStream(inputJarEntry);
+      String entryName = inputJarEntry.getName();
+      if (previousEntryNames.contains(entryName)) {
+        LOG.debug("Skipping duplicated file {}", entryName);
+      } else {
+        JarEntry outputJarEntry = new JarEntry(inputJarEntry);
+        outputStream.putNextEntry(outputJarEntry);
+        LOG.trace("Copying jar entry {}", inputJarEntry);
+        IOUtils.copy(inputStream, outputStream);
+        previousEntryNames.add(entryName);
+      }
+    }
+  }
+
+  private static void writePipeline(Pipeline pipeline, JarOutputStream outputStream)
+      throws IOException {
+    JarEntry jarEntry = new JarEntry(PortablePipelineJarUtils.PIPELINE_FILE_NAME);
+    outputStream.putNextEntry(jarEntry);
+    pipeline.writeTo(outputStream);
+  }
+
+  private static void writePipelineOptions(Struct optionsProto, JarOutputStream outputStream)
+      throws IOException {
+    JarEntry jarEntry = new JarEntry(PortablePipelineJarUtils.PIPELINE_OPTIONS_FILE_NAME);
+    outputStream.putNextEntry(jarEntry);
+    optionsProto.writeTo(outputStream);
+  }
+
+  @VisibleForTesting
+  interface ArtifactRetriever {
+    GetManifestResponse getManifest(GetManifestRequest request);
+
+    Iterator<ArtifactChunk> getArtifact(GetArtifactRequest request);
+  }
+
+  /**
+   * Copy all artifacts retrievable via the {@link ArtifactRetrievalServiceBlockingStub} to the
+   * {@code outputStream}.
+   *
+   * @return A {@link ProxyManifest} pointing to the artifacts' location in the output jar.
+   */
+  @VisibleForTesting
+  static ProxyManifest copyStagedArtifacts(
+      String retrievalToken, JarOutputStream outputStream, ArtifactRetriever retrievalServiceStub)
+      throws IOException {
+    GetManifestRequest manifestRequest =
+        GetManifestRequest.newBuilder().setRetrievalToken(retrievalToken).build();
+    ArtifactApi.Manifest manifest = retrievalServiceStub.getManifest(manifestRequest).getManifest();
```

Comment:
This require double copying of the artifacts. 1. To staging location 2. In to the jar.
This should be ok for now but will it be easy use artifact staging service to write directly to the jar.
Also, should we delete the temporarily staged artifacts?

Reply:
The JarArtifactStagingService approach is possible, but it would require a few non-trivial changes:

1. Abstract out the parts of [BeamFileSystemArtifactStagingService](https://github.com/apache/beam/blob/c2f0d282337f3ae0196a7717712396a5a41fdde1/runners/java-fn-execution/src/main/java/org/apache/beam/runners/fnexecution/artifact/BeamFileSystemArtifactStagingService.java) that write the artifacts.
2. Create a new writer that writes to a JarOutputStream instead of a file system.
3. Find some way to change the output jar name. Right now, output paths are generated automatically, and the artifact API doesn't pass any options to make this configurable.

The way things are now, the temporarily staged artifacts are deleted when `--clean-artifacts-per-job` is enabled.

Reply:
Ya, it will add additional complexity. Lets leave it as it is and enable the `--clean-artifacts-per-job` flag only when creating the jar.